rs/nns/governance/canister/canister.rs

```diff
@@ -131,27 +134,35 @@ are best practices for making use of the unsafe block:
 /// This should only be called once the global state has been initialized, which
 /// happens in `canister_init` or `canister_post_upgrade`.
 fn governance() -> &'static Governance {
-    unsafe { GOVERNANCE.as_ref().expect("Canister not initialized!") }
+    unsafe {
+        &*GOVERNANCE.with(|g| {
+            let ptr = g.as_ptr();
+            (*ptr).as_ref().expect("Canister not initialized!")
+        })
+    }
 }
 
 /// Returns a mutable reference to the global state.
 ///
 /// This should only be called once the global state has been initialized, which
 /// happens in `canister_init` or `canister_post_upgrade`.
 fn governance_mut() -> &'static mut Governance {
-    unsafe { GOVERNANCE.as_mut().expect("Canister not initialized!") }
+    unsafe {
```

Comment:
If we do need governance_mut(), we should instead return a `RefMut` wrapper which ensures that the code panics if someone attempts to create two mutable refs (or a shared refs + a mut ref) 
Similarly, `governance()` should return a `Ref`.

Reply:
I think the best way would be checking all the places which call methods with `&mut self` on `Governance`. I would then change these methods to work with `&self` by using interior mutability.

The other options would be to provide wrappers for `&mut self` functions, for example:

```
pub fn set_proposal_execution_status(pid: u64, result: Result<(), GovernanceError>) {
    // Note that this wrapper ensures that borrow_mut() is not held for longer than necessary.
    GOVERNANCE.borrow_mut().set_proposal_execution_status(pid, result)
}
```

Reply:
I mostly agree with what you said, and those actually are the main motivations for this PR. The reason we cannot get rid of UB right away is that we have a large amount of code depending on this bad access pattern (since genesis). This PR allows us to gradually move away from this pattern (I think 80% can be removed right away, but the remaining 20% needs a lot of refactoring of critical code)